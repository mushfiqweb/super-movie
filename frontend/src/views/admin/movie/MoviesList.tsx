import axios from 'axios';
import React, { useEffect, useState } from 'react';

type MovieItemType = {
    id: number;
    title: string;
    director: string;
    releaseYear: number;
    description: string;
    rating: number;
    duration: number;
    genres: string[];
    isAvailable: boolean;
    posterUrl: string;
    createdAt: Date;
    updatedAt: Date;
};

export const MovieList = () => {
    const [movies, setMovies] = useState<MovieItemType[]>([]);

    useEffect(() => {
        fetchMovies();
    }, []);

    const fetchMovies = async () => {
        try {
            const response = await axios.get<MovieItemType[]>('/movies');
            setMovies(response.data);
        } catch (error) {
            console.error('Error fetching movies:', error);
        }
    };

    return (
        <div>
            <h2>Movie List</h2>
            {movies.map((movie) => (
                <div key={movie.id}>
                    <h3>{movie.title}</h3>
                    <p>Director: {movie.director}</p>
                    <p>Release Year: {movie.releaseYear}</p>
                    {/* Add more movie details as needed */}
                </div>
            ))}
        </div>
    );
};

const MovieCreate = () => {
    const [newMovie, setNewMovie] = useState<MovieItemType>({
        id: 0,
        title: '',
        director: '',
        releaseYear: 0,
        description: '',
        rating: 0,
        duration: 0,
        genres: [],
        isAvailable: false,
        posterUrl: '',
        createdAt: new Date(),
        updatedAt: new Date(),
    });

    const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
        setNewMovie({ ...newMovie, [e.target.name]: e.target.value });
    };

    const handleGenreChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const genre = e.target.value;
        if (e.target.checked) {
            setNewMovie({ ...newMovie, genres: [...newMovie.genres, genre] });
        } else {
            setNewMovie({ ...newMovie, genres: newMovie.genres.filter((g) => g !== genre) });
        }
    };

    const handleSubmit = async (e: React.FormEvent) => {
        e.preventDefault();
        try {
            await axios.post('/movies', newMovie);
            // Reset form or perform any additional actions
        } catch (error) {
            console.error('Error creating movie:', error);
        }
    };

    return (
        <div>
            <h2>Create Movie</h2>
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="title"
                    placeholder="Title"
                    value={newMovie.title}
                    onChange={handleChange}
                />
                <input
                    type="text"
                    name="director"
                    placeholder="Director"
                    value={newMovie.director}
                    onChange={handleChange}
                />
                <input
                    type="number"
                    name="releaseYear"
                    placeholder="Release Year"
                    value={newMovie.releaseYear}
                    onChange={handleChange}
                />
                <textarea
                    name="description"
                    placeholder="Description"
                    value={newMovie.description}
                    onChange={handleChange}
                />
                <input
                    type="number"
                    name="rating"
                    placeholder="Rating"
                    value={newMovie.rating}
                    onChange={handleChange}
                />
                <input
                    type="number"
                    name="duration"
                    placeholder="Duration"
                    value={newMovie.duration}
                    onChange={handleChange}
                />
                <div>
                    <label>Genres:</label>
                    <div>
                        <input
                            type="checkbox"
                            name="genres"
                            value="Action"
                            checked={newMovie.genres.includes('Action')}
                            onChange={handleGenreChange}
                        />
                        <label>Action</label>
                    </div>
                    {/* Add more checkboxes for other genres */}
                </div>
                <input
                    type="checkbox"
                    name="isAvailable"
                    checked={newMovie.isAvailable}
                    onChange={(e) => setNewMovie({ ...newMovie, isAvailable: e.target.checked })}
                />
                <label>Available</label>
                <input
                    type="text"
                    name="posterUrl"
                    placeholder="Poster URL"
                    value={newMovie.posterUrl}
                    onChange={handleChange}
                />
                <button type="submit">Create Movie</button>
            </form>
        </div>
    );
};

const MovieUpdate = ({ movie }: { movie: MovieItemType }) => {
    const [updatedMovie, setUpdatedMovie] = useState<MovieItemType>(movie);

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setUpdatedMovie({ ...updatedMovie, [e.target.name]: e.target.value });
    };

    const handleSubmit = async (e: React.FormEvent) => {
        e.preventDefault();
        try {
            await axios.put(`/movies/${movie.id}`, updatedMovie);
            // Perform any additional actions after successful update
        } catch (error) {
            console.error('Error updating movie:', error);
        }
    };

    return (
        <div>
            <h2>Update Movie</h2>
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="title"
                    placeholder="Title"
                    value={updatedMovie.title}
                    onChange={handleChange}
                />
                {/* Add more input fields for other movie properties */}
                <button type="submit">Update Movie</button>
            </form>
        </div>
    );
};

const MovieDelete = ({ movieId }: { movieId: number }) => {
    const handleDelete = async () => {
        try {
            await axios.delete(`/movies/${movieId}`);
            // Perform any additional actions after successful deletion
        } catch (error) {
            console.error('Error deleting movie:', error);
        }
    };

    return (
        <div>
            <button onClick={handleDelete}>Delete Movie</button>
        </div>
    );
};
