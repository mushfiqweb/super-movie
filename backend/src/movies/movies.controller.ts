import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Put,
} from '@nestjs/common';
import { Movie } from './entities/movie.entity';
import { MoviesService } from './movies.service';

@Controller('movies')
export class MoviesController {
  constructor(private readonly moviesService: MoviesService) {}

  @Get()
  findAll() {
    return this.moviesService.findAll();
  }

  @Get(':id')
  findOne(@Param('id') id: number) {
    return this.moviesService.findOne(id);
  }

  @Post()
  create(@Body() movie: Movie) {
    console.log({ movie });
    return this.moviesService.create(movie);
  }

  @Put(':id')
  update(@Param('id') id: number, @Body() movie: Movie) {
    return this.moviesService.update(id, movie);
  }

  @Delete(':id')
  remove(@Param('id') id: number) {
    return this.moviesService.remove(id);
  }
}
