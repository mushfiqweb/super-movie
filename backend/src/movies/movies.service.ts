import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Movie } from './entities/movie.entity';

@Injectable()
export class MoviesService {
  constructor(
    @InjectRepository(Movie)
    private moviesRepository: Repository<Movie>,
  ) {}

  async findAll(): Promise<Movie[]> {
    return this.moviesRepository.find();
  }

  async findOne(id: number): Promise<Movie | undefined> {
    return this.moviesRepository.findOne({ where: { id } });
  }

  async create(movie: Movie): Promise<Movie> {
    return this.moviesRepository.save(movie);
  }

  async update(id: number, movie: Movie): Promise<Movie> {
    await this.moviesRepository.update(id, movie);
    return this.moviesRepository.findOne({ where: { id } });
  }

  async remove(id: number): Promise<void> {
    await this.moviesRepository.delete(id);
  }
}
