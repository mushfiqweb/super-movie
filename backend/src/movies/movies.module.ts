import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Movie } from './entities/movie.entity';
import { MoviesController } from './movies.controller';
import { MoviesService } from './movies.service';

@Module({
  imports: [TypeOrmModule.forFeature([Movie]), HttpModule],
  providers: [MoviesService],
  controllers: [MoviesController],
})
export class MoviesModule {}
