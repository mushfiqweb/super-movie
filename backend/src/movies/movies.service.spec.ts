import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { DeepPartial, Repository } from 'typeorm';
import { Movie } from './entities/movie.entity';
import { MoviesService } from './movies.service';

describe('MoviesService', () => {
  let service: MoviesService;
  let repository: Repository<Movie>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        MoviesService,
        {
          provide: getRepositoryToken(Movie),
          useClass: Repository,
        },
      ],
    }).compile();

    service = module.get<MoviesService>(MoviesService);
    repository = module.get<Repository<Movie>>(getRepositoryToken(Movie));
  });

  describe('findAll', () => {
    it('should return an array of movies', async () => {
      const mockMovies = [
        {
          id: 1,
          title: 'Movie 1',
          director: 'Director 1',
          releaseYear: 2022,
          description: 'Description 1',
          posterUrl: 'https://example.com/poster1.jpg',
          rating: 4.5,
          duration: 120,
          genres: ['Action', 'Adventure'],
          isAvailable: true,
        },
        {
          id: 2,
          title: 'Movie 2',
          director: 'Director 2',
          releaseYear: 2021,
          description: 'Description 2',
          posterUrl: 'https://example.com/poster2.jpg',
          rating: 3.8,
          duration: 90,
          genres: ['Comedy', 'Romance'],
          isAvailable: false,
        },
      ];

      jest.spyOn(repository, 'find').mockResolvedValue(mockMovies);

      const result = await service.findAll();

      expect(result).toEqual(mockMovies);
    });
  });

  describe('findOne', () => {
    it('should return a movie by id', async () => {
      const mockMovie = {
        id: 1,
        title: 'Movie 1',
        director: 'Director 1',
        releaseYear: 2022,
        description: 'Description 1',
        posterUrl: 'https://example.com/poster1.jpg',
        rating: 4.5,
        duration: 120,
        genres: ['Action', 'Adventure'],
        isAvailable: true,
      };

      jest.spyOn(repository, 'findOne').mockResolvedValue(mockMovie);

      const result = await service.findOne(1);

      expect(result).toEqual(mockMovie);
    });
  });

  describe('create', () => {
    it('should create a new movie', async () => {
      const mockMovie: DeepPartial<Movie> = {
        title: 'New Movie',
        director: 'New Director',
        releaseYear: 2023,
        description: 'New Description',
        posterUrl: 'https://example.com/new-poster.jpg',
        rating: 4.2,
        duration: 105,
        genres: ['Drama'],
        isAvailable: true,
      };

      const newMovie = new Movie();
      Object.assign(newMovie, mockMovie);

      jest.spyOn(repository, 'save').mockResolvedValue(newMovie);

      const result = await service.create(newMovie);

      expect(result).toEqual(newMovie);
    });
  });

  describe('update', () => {
    it('should update an existing movie', async () => {
      const mockMovie = {
        id: 1,
        title: 'Movie 1',
        director: 'Director 1',
        releaseYear: 2022,
        description: 'Description 1',
        posterUrl: 'https://example.com/poster1.jpg',
        rating: 4.5,
        duration: 120,
        genres: ['Action', 'Adventure'],
        isAvailable: true,
      };

      jest.spyOn(repository, 'update').mockResolvedValue(undefined);
      jest.spyOn(repository, 'findOne').mockResolvedValue(mockMovie);

      const result = await service.update(1, mockMovie);

      expect(result).toEqual(mockMovie);
    });
  });

  describe('remove', () => {
    it('should remove a movie by id', async () => {
      jest.spyOn(repository, 'delete').mockResolvedValue(undefined);

      await service.remove(1);

      expect(repository.delete).toHaveBeenCalledWith(1);
    });
  });

  describe('Snapshot tests', () => {
    it('should match the snapshot for findAll', async () => {
      const mockMovies = [
        {
          id: 1,
          title: 'Movie 1',
          director: 'Director 1',
          releaseYear: 2022,
          description: 'Description 1',
          posterUrl: 'https://example.com/poster1.jpg',
          rating: 4.5,
          duration: 120,
          genres: ['Action', 'Adventure'],
          isAvailable: true,
        },
        {
          id: 2,
          title: 'Movie 2',
          director: 'Director 2',
          releaseYear: 2021,
          description: 'Description 2',
          posterUrl: 'https://example.com/poster2.jpg',
          rating: 3.8,
          duration: 90,
          genres: ['Comedy', 'Romance'],
          isAvailable: false,
        },
      ];

      jest.spyOn(repository, 'find').mockResolvedValue(mockMovies);

      const result = await service.findAll();

      expect(result).toMatchSnapshot();
    });

    it('should match the snapshot for findOne', async () => {
      const mockMovie = {
        id: 1,
        title: 'Movie 1',
        director: 'Director 1',
        releaseYear: 2022,
        description: 'Description 1',
        posterUrl: 'https://example.com/poster1.jpg',
        rating: 4.5,
        duration: 120,
        genres: ['Action', 'Adventure'],
        isAvailable: true,
      };

      jest.spyOn(repository, 'findOne').mockResolvedValue(mockMovie);

      const result = await service.findOne(1);

      expect(result).toMatchSnapshot();
    });
  });
});
