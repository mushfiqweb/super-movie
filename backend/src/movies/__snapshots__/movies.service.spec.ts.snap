// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesService Snapshot tests should match the snapshot for findAll 1`] = `
[
  {
    "description": "Description 1",
    "director": "Director 1",
    "duration": 120,
    "genres": [
      "Action",
      "Adventure",
    ],
    "id": 1,
    "isAvailable": true,
    "posterUrl": "https://example.com/poster1.jpg",
    "rating": 4.5,
    "releaseYear": 2022,
    "title": "Movie 1",
  },
  {
    "description": "Description 2",
    "director": "Director 2",
    "duration": 90,
    "genres": [
      "Comedy",
      "Romance",
    ],
    "id": 2,
    "isAvailable": false,
    "posterUrl": "https://example.com/poster2.jpg",
    "rating": 3.8,
    "releaseYear": 2021,
    "title": "Movie 2",
  },
]
`;

exports[`MoviesService Snapshot tests should match the snapshot for findOne 1`] = `
{
  "description": "Description 1",
  "director": "Director 1",
  "duration": 120,
  "genres": [
    "Action",
    "Adventure",
  ],
  "id": 1,
  "isAvailable": true,
  "posterUrl": "https://example.com/poster1.jpg",
  "rating": 4.5,
  "releaseYear": 2022,
  "title": "Movie 1",
}
`;
